pkg/runtime/controller.go

```diff
@@ -0,0 +1,424 @@
+type rulesListByNamespace map[string][]*Rule
+
+// convertToRuntimeRules converts internal rules to the format that resolver needs.
+func convertToRuntimeRules(ruleConfig rulesMapByNamespace) (rulesListByNamespace, int) {
+	// convert rules
+	nrules := 0
+	rules := make(rulesListByNamespace)
+	for ns, nsmap := range ruleConfig {
+		rulesArr := make([]*Rule, 0, len(nsmap))
+		for _, rule := range nsmap {
+			rulesArr = append(rulesArr, rule)
+		}
+		rules[ns] = rulesArr
+		nrules += len(rulesArr)
+	}
+	return rules, nrules
+}
+
+// processRules builds the current consistent view of the rules keyed by Namespace and then Name.
+// ht (handlerTable) keeps track of handler-instance association.
+func (c *Controller) processRules(handlerConfig map[string]*cpb.Handler,
+	instanceConfig map[string]*cpb.Instance, ht *handlerTable) rulesMapByNamespace {
+	// current consistent view of the rules
+	// keyed by Namespace and then Name.
+	ruleConfig := make(rulesMapByNamespace)
+
+	// check rules and ensure only good handlers and instances are used.
+	// record handler - instance associations
+	for k, cfg := range c.configState {
+		if k.Kind != RulesKind {
+			continue
+		}
+		rulec := cfg.(*cpb.Rule)
+		rule := &Rule{
+			selector: rulec.Selector,
+			name:     k.Name,
+		}
+		acts := c.processActions(rulec.Actions, handlerConfig, instanceConfig, ht)
+
+		rn := ruleConfig[k.Namespace]
+		if rn == nil {
+			rn = make(map[string]*Rule)
+			ruleConfig[k.Namespace] = rn
+		}
+		ruleActions := make(map[adptTmpl.TemplateVariety][]*Action)
+		for vr, amap := range acts {
+			for _, cf := range amap {
+				ruleActions[vr] = append(ruleActions[vr], cf)
+			}
+		}
+		rule.actions = ruleActions
+		rn[k.Name] = rule
+	}
+
+	return ruleConfig
+}
+
+var cleanupSleepTime = 500 * time.Millisecond
+
+// processActions prunes actions that lack referential integrity and associate instances with
+// handlers that are later used to create new handlers.
+func (c *Controller) processActions(acts []*cpb.Action, handlerConfig map[string]*cpb.Handler,
+	instanceConfig map[string]*cpb.Instance, ht *handlerTable) map[adptTmpl.TemplateVariety]map[string]*Action {
+	actions := make(map[adptTmpl.TemplateVariety]map[string]*Action)
+	for _, ic := range acts {
+		var hc *cpb.Handler
+		if hc = handlerConfig[ic.Handler]; hc == nil {
+			if glog.V(3) {
+				glog.Warningf("ConfigWarning unknown handler: %s", ic.Handler)
+			}
+			continue
+		}
+
+		for _, instName := range ic.Instances {
+			var inst *cpb.Instance
+			if inst = instanceConfig[instName]; inst == nil {
+				if glog.V(3) {
+					glog.Warningf("ConfigWarning unknown instance: %s", instName)
+				}
+				continue
+			}
+
+			ht.Associate(ic.Handler, instName)
+
+			ti := c.templateInfo[inst.Template]
+			// grouped by adapterName.
+			vAction := actions[ti.Variety]
+			if vAction == nil {
+				vAction = make(map[string]*Action)
+				actions[ti.Variety] = vAction
+			}
+
+			act := vAction[hc.Adapter]
```

Comment:
I am not sure I understand : Why are we merging instances across multiple actions that go to the same adapter.. It seems we are only creating one Action object for all the actions that have the same templateVariety and same AdapterName and merging their instances.. I might be missing something, but from the top level it seems incorrect.